test/cpp/end2end/spiffe_bundle_map_end2end_test.cc

```diff
@@ -0,0 +1,474 @@
+constexpr absl::string_view kServerChainCertPath =
+    "test/core/tsi/test_creds/spiffe_end2end/leaf_and_intermediate_chain.pem";
+constexpr absl::string_view kClientSpiffeBundleMapPath =
+    "test/core/tsi/test_creds/spiffe_end2end/client_spiffebundle.json";
+constexpr absl::string_view kServerSpiffeBundleMapPath =
+    "test/core/tsi/test_creds/spiffe_end2end/server_spiffebundle.json";
+
+std::string MakeConnectionFailureRegex(absl::string_view prefix) {
+  return absl::StrCat(
+      prefix,
+      "(UNKNOWN|UNAVAILABLE): "
+      // IP address
+      "(ipv6:%5B::1%5D|ipv4:127.0.0.1):[0-9]+: "
+      // Prefixes added for context
+      "(Failed to connect to remote host: )?"
+      "(Timeout occurred: )?"
+      // Parenthetical wrappers
+      "( ?\\(*("
+      "Secure read failed|"
+      "Handshake (read|write) failed|"
+      "Delayed close due to in-progress write|"
+      // Syscall
+      "((connect|sendmsg|recvmsg|getsockopt\\(SO\\_ERROR\\)): ?)?"
+      // strerror() output or other message
+      "(Connection refused"
+      "|Connection reset by peer"
+      "|Socket closed"
+      "|Broken pipe"
+      "|FD shutdown"
+      "|Endpoint closing)"
+      // errno value
+      "( \\([0-9]+\\))?"
+      // close paren from wrappers above
+      ")\\)*)+");
+}
+
+std::string MakeTlsHandshakeFailureRegex(absl::string_view prefix) {
+  return absl::StrCat(
+      prefix,
+      "(UNKNOWN|UNAVAILABLE): "
+      // IP address
+      "(ipv6:%5B::1%5D|ipv4:127.0.0.1):[0-9]+: "
+      // Prefixes added for context
+      "(Failed to connect to remote host: )?"
+      // Tls handshake failure
+      "Tls handshake failed \\(TSI_PROTOCOL_FAILURE\\): SSL_ERROR_SSL: "
+      "error:1000007d:SSL routines:OPENSSL_internal:CERTIFICATE_VERIFY_FAILED"
+      // Detailed reason for certificate verify failure
+      "(: .*)?");
+}
+
+class SpiffeBundleMapTest : public ::testing::Test {
+ protected:
+  void RunServer(absl::Notification* notification, absl::string_view key_path,
+                 absl::string_view cert_path, absl::string_view root_path,
+                 absl::string_view spiffe_bundle_map_path) {
+    auto certificate_provider =
+        std::make_shared<experimental::FileWatcherCertificateProvider>(
+            std::string(key_path), std::string(cert_path),
+            std::string(root_path), std::string(spiffe_bundle_map_path), 1);
+    grpc::experimental::TlsServerCredentialsOptions options(
+        certificate_provider);
+    options.watch_root_certs();
+    options.set_root_cert_name("root");
+    options.watch_identity_key_cert_pairs();
+    options.set_identity_cert_name("identity");
+    options.set_cert_request_type(
+        GRPC_SSL_REQUEST_AND_REQUIRE_CLIENT_CERTIFICATE_AND_VERIFY);
+    auto server_credentials = grpc::experimental::TlsServerCredentials(options);
+    CHECK_NE(server_credentials.get(), nullptr);
+
+    grpc::ServerBuilder builder;
+    TestServiceImpl service_;
```

Comment:
We seem to be redeclaring `service_` when it is already a member field of the class - is this a bug?